File: libs/java/src/main/java/com/breissinger/ggcommons/messaging/providers/greengrass/SubscriptionHandler.java
/*
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
package com.breissinger.ggcommons.messaging.providers.greengrass;


import com.breissinger.ggcommons.messaging.Message;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import oshi.util.tuples.Pair;
import software.amazon.awssdk.eventstreamrpc.StreamResponseHandler;

import java.util.concurrent.*;
import java.util.function.BiConsumer;

public abstract class SubscriptionHandler<T> implements Runnable, StreamResponseHandler<T>
{
    protected static final Logger LOGGER = LogManager.getLogger(SubscriptionHandler.class);

    protected static class QueueEntry
    {
        public String topic;
        public Message message;

        QueueEntry(String topic, Message message)
        {
            this.topic = topic;
            this.message = message;
        }
    }

    protected String topicFilter;
    protected BiConsumer<String, Message> callback;
    protected int maxConcurrency;
    protected int maxMessages;
    LinkedBlockingQueue<QueueEntry> queue;
    ExecutorService executor;
    private final Semaphore concurrencyLimit;
    private volatile boolean shutdown = false;

    public SubscriptionHandler(String topicFilter, BiConsumer<String, Message> callback, int maxConcurrency, int maxMessages)
    {
        this.topicFilter = topicFilter;
        this.callback = callback;
        this.maxConcurrency = maxConcurrency;
        this.maxMessages = maxMessages;
        // Bounded queue (drop on overflow) when maxMessages > 0, else unbounded — parity with
        // the Rust (bounded mpsc) / TS (drop-on-overflow) providers.
        this.queue = maxMessages > 0 ? new LinkedBlockingQueue<>(maxMessages) : new LinkedBlockingQueue<>();
        // One virtual thread per callback (callbacks block on IPC / IoT Core / CloudWatch
        // I/O). A positive maxConcurrency is enforced with a Semaphore, preserving the
        // bounded-concurrency contract without a fixed platform-thread pool.
        executor = Executors.newVirtualThreadPerTaskExecutor();
        concurrencyLimit = maxConcurrency > 0 ? new Semaphore(maxConcurrency) : null;
        new Thread(this, "gg-sub-" + topicFilter).start();
    }

    public void shutdown() {
        shutdown = true;
        queue.offer(new QueueEntry(null, null)); // Poison pill
        executor.shutdown();
        try {
            if (!executor.awaitTermination(5, TimeUnit.SECONDS)) {
                executor.shutdownNow();
            }
        } catch (InterruptedException e) {
            executor.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }
    
    abstract Pair<String,Message> parseRawPayload(T rawPayload);

    @Override
    public void onStreamEvent(T rawMessage)
    {
        Pair<String, Message> parsedMessage = parseRawPayload(rawMessage);
        if (parsedMessage != null)
        {
            // Non-blocking enqueue: a full bounded queue drops the message with a warning rather
            // than blocking the IPC stream thread (parity with Rust/TS).
            if (!queue.offer(new QueueEntry(parsedMessage.getA(), parsedMessage.getB())))
            {
                LOGGER.warn("Subscription queue full (maxMessages={}) for '{}'; dropping message on {}",
                        maxMessages, topicFilter, parsedMessage.getA());
            }
        }
    }

    @Override
    public boolean onStreamError(Throwable throwable)
    {
        LOGGER.error("Error on stream for subscription to topicFilter {}: {}", topicFilter, throwable.toString());
        return false;
    }

    @Override
    public void onStreamClosed()
    {
        LOGGER.info("Stream for subscription to topicFilter {} closed (unsubscribed)", topicFilter);
        shutdown();
    }

    @Override
    public void run()
    {
        LOGGER.info("Starting queue monitoring for subscription on {}", topicFilter);
        while(!shutdown) {
            try
            {
                final QueueEntry entry = queue.take();
                if (entry.message == null && entry.topic == null)
                    break;
                if (concurrencyLimit != null)
                {
                    // Backpressure: block the drain thread until a permit frees, so at
                    // most maxConcurrency callbacks run (and are in flight) at once.
                    concurrencyLimit.acquireUninterruptibly();
                }
                executor.execute(() -> {
                    try
                    {
                        LOGGER.trace("Invoking callback for topic '{}'", entry.topic);
                        callback.accept(entry.topic, entry.message);
                    }
                    catch (Throwable t)
                    {
                        // A subscription callback must never throw out of this task: an uncaught
                        // exception here kills the worker thread and (under crash/restart churn) can
                        // surface in the eventstream RPC layer and destabilize nucleus IPC. Contain
                        // and log it so one bad message can't take down the subscription or the core.
                        LOGGER.error("Subscription callback for topic '{}' threw; suppressed: {}",
                                entry.topic, t.toString(), t);
                    }
                    finally
                    {
                        if (concurrencyLimit != null)
                        {
                            concurrencyLimit.release();
                        }
                    }
                });
            }
            catch (InterruptedException e)
            {
                LOGGER.warn("Subscription processing for {} interrupted. Restarting. Exception: {}",
                        topicFilter, e.getMessage());
            }
        }
        LOGGER.info("Queue monitoring stopped for subscription on {}", topicFilter);
    }
}

File: libs/python/edgecommons/messaging/providers/greengrass/subscription_handler.py
import abc
import concurrent.futures.thread
import logging
import queue
from threading import Lock, Thread, current_thread
from typing import Callable
from edgecommons.messaging.message import Message
from edgecommons.messaging.messaging_provider import DEFAULT_MAX_MESSAGES


logger = logging.getLogger("SubscriptionHandler")


class SubscriptionHandler(metaclass=abc.ABCMeta):
    _STOP = object()

    def __init__(
        self,
        topic_filter,
        callback: Callable[[str, Message], None],
        max_concurrency: int = None,
        max_messages: int = None,
    ):
        self._topic_filter = topic_filter
        self._callback_func = callback
        # Bounded queue (drop on overflow) when max_messages > 0, else unbounded — parity with
        # the Rust (bounded mpsc) / TS (drop-on-overflow) providers.
        self._max_messages = max_messages if max_messages is not None else DEFAULT_MAX_MESSAGES
        self._queue = (
            queue.Queue(maxsize=self._max_messages)
            if self._max_messages and self._max_messages > 0
            else queue.Queue()
        )
        self._max_concurrency = max_concurrency
        self._closed = False
        self._close_lock = Lock()
        self._thread = Thread(target=self._process_queue, daemon=True)
        self._thread.start()

    def _invoke_callback(self, topic: str, msg: Message) -> None:
        # Wrap the user/lib callback so an exception it throws can never escape the
        # worker thread. An uncaught callback error on this thread would otherwise
        # surface in the eventstream RPC layer and can wedge the nucleus's single IPC
        # event-loop thread under crash/restart churn. Log and suppress instead.
        try:
            self._callback_func(topic, msg)
        except Exception as error:  # noqa: BLE001 - never let a bad message kill the worker
            logger.error(
                f"Exception {error} in subscription callback for topic '{topic}'"
                f" (filter '{self._topic_filter}'); suppressing to keep the worker alive"
            )

    def get_topic_filter(self):
        return self._topic_filter

    def get_max_concurrency(self):
        return self._max_concurrency

    @abc.abstractmethod
    def parse_raw_payload(self, event):
        pass

    def on_stream_error(self, error: Exception) -> bool:
        logger.error(
            f"Stream error: {error} for topic filter {self._topic_filter}. Closing stream."
        )
        return True  # Return True to close stream, False to keep stream open.

    def on_stream_closed(self) -> None:
        self.close()
        logger.debug(f"Subscription stream on topic {self._topic_filter} closed")

    def close(self, timeout: float = 2.0) -> None:
        with self._close_lock:
            if self._closed:
                return
            self._closed = True
        try:
            self._queue.put_nowait(self._STOP)
        except queue.Full:
            try:
                self._queue.get_nowait()
            except queue.Empty:
                pass
            try:
                self._queue.put_nowait(self._STOP)
            except queue.Full:
                logger.warning(
                    f"Could not enqueue close marker for subscription worker on "
                    f"'{self._topic_filter}'"
                )
        if current_thread() is not self._thread:
            self._thread.join(timeout=timeout)

    def on_stream_event(self, event) -> None:
        if self._closed:
            return
        logger.debug(f"Received stream message on topic filter: {self._topic_filter}")
        try:
            parsed = self.parse_raw_payload(event)
            if parsed is None:
                return
            topic, received_payload = parsed
            topic_payload_tuple = (topic, received_payload)
            # Non-blocking enqueue: a full bounded queue drops the message with a warning rather
            # than blocking the IPC stream thread (parity with Rust/TS).
            try:
                self._queue.put_nowait(topic_payload_tuple)
            except queue.Full:
                logger.warning(
                    f"subscription queue full (max_messages={self._max_messages}) for filter "
                    f"'{self._topic_filter}'; dropping message on {topic}"
                )
            logger.debug(
                f"IPC: common: PubSubDataHandler: on_stream_event: subscribed message: {received_payload}"
            )
        except Exception as error:
            logger.error(
                f"Exception {error} decoding payload for topic filter {self._topic_filter}"
            )
            logger.error(
                "Probable cause: payload is not an EdgeCommons protobuf message"
            )

    def _process_queue(self):
        logger.info(
            f"Starting queue monitoring for subscription on topic {self._topic_filter}"
        )
        with concurrent.futures.ThreadPoolExecutor(
            max_workers=self._max_concurrency
        ) as executor:
            while True:
                try:
                    queue_obj = self._queue.get()
                    if queue_obj is self._STOP:
                        break
                    topic = queue_obj[0]
                    msg = queue_obj[1]
                    executor.submit(self._invoke_callback, topic, msg)
                except Exception as e:
                    logger.warning(
                        f"Exception while processing message from subscription to '{self._topic_filter}': {e}"
                    )
        logger.info(
            f"Queue monitoring stopped for subscription on topic {self._topic_filter}"
        )
